### PR TITLE
AZ-411: Add rate limiting to the tx send

### PR DESCRIPTION
**Pull request by kostekIV**

### CODE REVIEW

File: flooder/src/config.rs [hunk 1]
```diff
@@ -39,6 +39,14 @@ pub struct Config {
     /// changes the awaited status of every transaction from Ready to SubmitOnly
     #[clap(long)]
     pub submit_only: bool,
+
+    /// How many transactions to put in the intervalt
```
Comment (fixxxedpoint) [inline flooder/src/config.rs:43]: typo

File: flooder/src/main.rs [hunk 2]
```diff
@@ -45,6 +41,12 @@ fn main() -> Result<(), anyhow::Error> {
             None => panic!("Needs --phrase or --seed"),
         },
     };
+
+    let rate_limiting = match (config.transactions_in_interval, config.interval_secs) {
```
Comment (fixxxedpoint) [inline flooder/src/main.rs:45]: Output something for the user. `Config.rs` doesn't provide any info on how to use these parameters together.

File: flooder/src/main.rs [hunk 3]
```diff
@@ -146,18 +148,37 @@ fn flood(
...
-        println!("Sending a batch of {} transactions", &batch.len());
-        batch.iter().enumerate().for_each(|(index, tx)| {
-            send_tx(
-                pool.get(index % pool.len()).unwrap(),
-                tx,
-                status,
-                Arc::clone(histogram),
-            )
-        })
+    let (transactions_in_interval, interval_dr) = rate_limit
```
Comment (fixxxedpoint) [inline flooder/src/main.rs:153]: Catch some free letters ;) `u a t i o n`

File: flooder/src/main.rs [hunk 4]
```diff
@@ -146,18 +148,37 @@ fn flood(
...
+        .map_or((txs.len(), Duration::from_secs(0)),
```
Comment (fixxxedpoint) [inline flooder/src/main.rs:154]: Are these defaults going to flood the log with messages from line 180? If so, refactor it slightly.

**Review by fixxxedpoint — CHANGES_REQUESTED** [review]

File: flooder/src/config.rs (hunk 1 above)
Comment (kostekIV) [reply flooder/src/config.rs]: fixed

File: flooder/src/main.rs (hunk 2 above)
Comment (kostekIV) [reply flooder/src/main.rs:45]: done

File: flooder/src/main.rs (hunk 3 above)
Comment (kostekIV) [reply flooder/src/main.rs]: thanks

File: flooder/src/main.rs (hunk 4 above)
Comment (kostekIV) [reply flooder/src/main.rs:155]: Only one message will appear

File: flooder/src/main.rs [hunk 5]
```diff
@@ -45,6 +41,13 @@ fn main() -> Result<(), anyhow::Error> {
             None => panic!("Needs --phrase or --seed"),
         },
     };
+
+    let rate_limiting = match (config.transactions_in_interval, config.interval_secs) {
+        (Some(tii), Some(is)) => Some((tii, is)),
+        (None, None) => None,
+        _ => panic!("--transactions-in-interval needs to be specified with--interval-secs"),
```
Comment (mike1729) [inline flooder/src/main.rs:48]: Nit: space misssing `with --interval`

File: flooder/src/main.rs [hunk 6]
```diff
@@ -146,18 +149,37 @@ fn flood(
...
+                |(transactions_in_interval, secs)| {
+                    (transactions_in_interval as usize, Duration::from_secs(secs))
+                });
+    let transactions_per_batch = transactions_in_interval / num_threads;
+
+    txs.chunks(transactions_in_interval).enumerate().for_each(|(interval_idx, interval)| {
+        let start = Instant::now();
+        info!("Starting {} interval", interval_idx);
+        interval.par_chunks(transactions_per_batch).for_each(|batch| {
+            println!("Sending a batch of {} transactions", &batch.len());
```
Comment (mike1729) [inline flooder/src/main.rs:165]: Replace with `info` or `debug`.

File: flooder/src/main.rs [hunk 7]
```diff
@@ -146,18 +149,37 @@ fn flood(
...
+                    tx,
+                    status,
+                    Arc::clone(histogram),
+                )
+            });
+        });
+        let exec_time = start.elapsed();
+
+        if let Some(remaining_time) = interval_duration.checked_sub(exec_time) {
+            info!("Sleeping for {}ms", remaining_time.as_millis());
```
Comment (mike1729) [inline flooder/src/main.rs:178]: Probably `debug` is more appropriate?

**Review by mike1729 — APPROVED** [review]

File: flooder/src/main.rs (hunk 7 above)
Comment (kostekIV) [reply flooder/src/main.rs]: done

File: flooder/src/main.rs (hunk 6 above)
Comment (kostekIV) [reply flooder/src/main.rs]: done

File: flooder/src/main.rs (hunk 5 above)
Comment (kostekIV) [reply flooder/src/main.rs]: done

**Pull request merged**